core/src/main/java/org/apache/lucene/analysis/miscellaneous/DuplicateByteSequenceSpotter.java

```diff
@@ -0,0 +1,325 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.lucene.analysis.miscellaneous;
+
+import org.apache.lucene.util.RamUsageEstimator;
+
+/**
+ * A Trie structure for analysing byte streams for duplicate sequences. Bytes
+ * from a stream are added one at a time using the addByte method and the number
+ * of times it has been seen as part of a sequence is returned.
+ * 
+ * The minimum required length for a duplicate sequence detected is 6 bytes.
+ * 
+ * The design goals are to maximize speed of lookup while minimizing the space
+ * required to do so. This has led to a hybrid solution for representing the
+ * bytes that make up a sequence in the trie.
+ * 
+ * If we have 6 bytes in sequence e.g. abcdef then they are represented as
+ * object nodes in the tree as follows:
+ * <p>
+ * (a)-(b)-(c)-(def as an int)
+ * <p>
+ * 
+ * 
+ * {@link RootTreeNode} objects are used for the first two levels of the tree
+ * (representing bytes a and b in the example sequence). The combinations of
+ * objects at these 2 levels are few so internally these objects allocate an
+ * array of 256 child node objects to quickly address children by indexing
+ * directly into the densely packed array using a byte value. The third level in
+ * the tree holds {@link LightweightTreeNode} nodes that have few children
+ * (typically much less than 256) and so use a dynamically-grown array to hold
+ * child objects of the type {@link ByteSequenceLeafNode}. These represent the
+ * final 3 bytes of a sequence as single int and hold a count of the number of
+ * times the entire sequence path has been visited.
+ * <p>
+ * The Trie grows indefinitely as more content is added and while theoretically
+ * it could be massive (a 6-depth tree could produce 256^6 nodes) non-random
+ * content e.g English text contains fewer variations.
+ * <p>
+ * In future we may look at using one of these strategies when memory is tight:
+ * <ol>
+ * <li>auto-pruning methods to remove less-visited parts of the tree
+ * <li>auto-reset to wipe the whole tree and restart when a memory threshold is
+ * reached
+ * <li>halting any growth of the tree
```

Comment:
DuplicateByteSequencer could have a future role/life in Lucene so I wanted to record memory usage here but not tie into elasticsearch